pkg/scheduler/plugins/reservation/transformer.go

```diff
@@ -22,210 +22,361 @@ import (
+	// multiple reusable Reservations that can be used by different Owners for the Pod and bypasses the Filter,
+	// we choose the instance with the earliest creation time.
+	var earliestReusableRInfo *reservationInfo
+	for _, v := range reservationInfos {
+		if !v.reservation.Spec.AllocateOnce {
+			if earliestReusableRInfo == nil ||
+				v.reservation.CreationTimestamp.Before(&earliestReusableRInfo.reservation.CreationTimestamp) {
+				earliestReusableRInfo = v
+				break
+			}
+		}
+	}
+
+	for _, rInfo := range reservationInfos {
+		if !rInfo.reservation.Spec.AllocateOnce && earliestReusableRInfo != nil && rInfo != earliestReusableRInfo {
+			continue
+		}
+		if err := restoreReservedResources(pl.handle, cycleState, pod, rInfo, nodeInfo, podInfoMap, shouldRestoreStates); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func restoreReservedResources(handle frameworkext.ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod, rInfo *reservationInfo, nodeInfo *framework.NodeInfo, podInfoMap map[types.UID]*framework.PodInfo, shouldRestoreStates bool) error {
+	reservePod := reservationutil.NewReservePod(rInfo.reservation)
+	reservePodInfo := framework.NewPodInfo(reservePod)
+
+	// Retain ports that are not used by other Pods. These ports need to be erased from NodeInfo.UsedPorts,
+	// otherwise it may cause Pod port conflicts
+	retainReservePodUnusedPorts(reservePod, rInfo.reservation, podInfoMap)
+
+	// When AllocateOnce is disabled, some resources may have been allocated,
+	// and an additional resource record will be accumulated at this time.
+	// Even if the Reservation is not bound by the Pod (e.g. Reservation is enabled with AllocateOnce),
+	// these resources held by the Reservation need to be returned, so as to ensure that
+	// the Pod can pass through each filter plugin during scheduling.
+	// The returned resources include scalar resources such as CPU/Memory, ports etc..
+	if err := nodeInfo.RemovePod(reservePod); err != nil {
+		return err
+	}
+	// NOTE: To achieve incremental update with frameworkext.TemporarySnapshot, we need to set Generation to -1
+	nodeInfo.Generation = -1
+
+	if !shouldRestoreStates {
+		return nil
+	}
+
+	// Regardless of whether the Reservation enables AllocateOnce
+	// and whether the Reservation uses high-availability constraints such as InterPodAffinity/PodTopologySpread,
+	// we should trigger the plugins to update the state and erase the Reservation-related state in this round
+	// of scheduling to ensure that the Pod can pass these filters.
+	// Although Reservation may reserve a large number of resources for a batch of Pods to use,
+	// and it is possible to use these high-availability constraints to cause resources to be unallocated and wasteful,
+	// users need to bear this waste.
+	status := handle.RunPreFilterExtensionRemovePod(context.Background(), cycleState, pod, reservePodInfo, nodeInfo)
```

Comment:
the core points of the plugin cache returning logic of matchedReservation are RemoveReservePod 、RemoveReservation、AddPodInReservation. What's the difference？Why not just call RemoveReservePod.